### PR TITLE
Iteratively run the main simplification pipeline.

### PR DESCRIPTION
**Pull request by silvasean**
(ignore the "Rework how global slot initializers work." commit -- this one doesn't trivially commute with it, but is still logically separate)

This introduces a new pass LowerToBackendContract (better name very
welcome) which performs the bulk of the simplifications that we do,
such as
- shape refinement
- dtype refinement
- maximizing value semantics
- inlining global slots
- decomposing complex ops

The key difference from before is that it iterates the set of
transformations, which can help to break a number of "catch-22" issues
where one simplification depends on another, the latest example being
here:
https://github.com/llvm/torch-mlir/issues/1131

### CODE REVIEW

**Comment by silvasean** [comment]
@ramiro050 PTAL now that I fixed the CI issue.

**Comment by silvasean** [comment]
+ @ashay for a second pair of eyes on this architectural change.

File: lib/Dialect/Torch/Transforms/Passes.cpp [hunk 1]
```diff
@@ -66,131 +70,82 @@ void mlir::torch::Torch::createTorchScriptModuleToTorchBackendPipeline(
...
   // Convert the bulk of non-ABI-visible !torch.tensor's to !torch.vtensor's.
   pm.addNestedPass<func::FuncOp>(Torch::createMaximizeValueSemanticsPass());
-
-  // Update the return op to return value tensors and remove dead ops.
+  // Update the return op to return value tensors.
   pm.addPass(Torch::createRefinePublicReturnPass());
   pm.addNestedPass<func::FuncOp>(createCanonicalizerPass());
-
-  // Ensure that all tensors have been converted to value semantics.
-  pm.addPass(Torch::createVerifyConversionToValueSemanticsPass());
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/Passes.cpp:153]: Is this pass no longer needed?

**Review by ramiro050 — APPROVED** [review]
This is great! I just have one question

File: lib/Dialect/Torch/Transforms/Passes.cpp (hunk 1 above)
Comment (silvasean) [reply lib/Dialect/Torch/Transforms/Passes.cpp:153]: It is all rolled into the termination condition of the LowerToBackendContract pass.

Comment (ramiro050) [reply lib/Dialect/Torch/Transforms/Passes.cpp:153]: Should the code for that pass be removed?

Comment (ramiro050) [reply lib/Dialect/Torch/Transforms/Passes.cpp:153]: Namely, https://github.com/silvasean/torch-mlir/blob/iterate/lib/Dialect/Torch/Transforms/VerifyConversionToValueSemantics.cpp

Comment (silvasean) [reply lib/Dialect/Torch/Transforms/Passes.cpp:153]: Oh, good call!

**Pull request merged**